### PR TITLE
Improve `apply_lifting` performance

### PR DESCRIPTION
**Pull request by jorgensd**
The current `apply_lifting` keeps on recomputing what cells has Dirichlet dofs in them.
This PR rewrites this such that we precompute this at the creation of the DirichletBC.
MWE.
```python
import dolfinx
from mpi4py import MPI
import time
import ufl
import numpy as np

import os
import os.path

Ns = [25, 50, 75]
num_runs = 10
name = "results.csv"

file_exists = os.path.exists(f"{os.getcwd()}/{name}")
if file_exists:
    outfile = open(name, "a")
else:
    outfile = open(name, "w")
    print("Hash Dofs Time", file=outfile)

for N in Ns:
    mesh = dolfinx.mesh.create_unit_cube(MPI.COMM_WORLD, N, N, N)

    V = dolfinx.fem.FunctionSpace(mesh, ("Lagrange", 1))
    u = ufl.TrialFunction(V)
    v = ufl.TestFunction(V)
    f = dolfinx.fem.Function(V)
    f.interpolate(lambda x: np.sin(x[0]))

    u_bc = dolfinx.fem.Function(V)
    u_bc.interpolate(lambda x: np.ones(x.shape[1])+np.cos(x[0]))
    mesh.topology.create_connectivity(mesh.topology.dim-1, mesh.topology.dim)
    exterior_facets = dolfinx.mesh.exterior_facet_indices(mesh.topology)
    dofs = dolfinx.fem.locate_dofs_topological(
        V, mesh.topology.dim-1, exterior_facets)

    bc = dolfinx.fem.dirichletbc(u_bc, dofs)

    a = ufl.inner(u, v) * ufl.dx
    L = ufl.inner(f, v) * ufl.dx

    af = dolfinx.fem.form(
        a,  jit_options={"cffi_extra_compile_args": ["-march=native", "-O3"]})
    Lf = dolfinx.fem.form(
        L, jit_options={"cffi_extra_compile_args": ["-march=native", "-O3"]})

    A = dolfinx.fem.petsc.assemble_matrix(af, bcs=[bc])
    start = time.perf_counter()
    b = dolfinx.fem.petsc.assemble_vector(Lf)
    end = time.perf_counter()

    num_dofs = V.dofmap.index_map.size_global * V.dofmap.index_map_bs

    for i in range(num_runs):
        start = time.perf_counter()
        dolfinx.fem.petsc.apply_lifting(b, [af], [[bc]])
        end = time.perf_counter()
        print(f"{dolfinx.git_commit_hash} {num_dofs} {end-start}", file=outfile)

outfile.close()
```
with plotting function:
```python
import matplotlib.pyplot as plt
import seaborn
import pandas
df = pandas.read_csv("results.csv", delimiter=" ")
embed()

plot = seaborn.catplot(data=df, kind="swarm", x="Dofs",
                       y="Time", hue="Hash")
plot.set(yscale="log")
plt.grid()
plt.savefig("runtimes.png")
```
yielding the following speed-up (in serial)
runtimes
where 9eee1fff652448e507b87b69f089200171160c0a is this PR  and ac414e45eb003a6610a180b3709fc5f743976e8d is main at time of writing

### CODE REVIEW

File: cpp/dolfinx/fem/DirichletBC.h [hunk 1]
```diff
@@ -131,6 +131,62 @@ template <typename T>
...
+    for (std::size_t i = 0; i < num_cells; ++i)
+    {
+      auto dofs = dofmap->cell_dofs(i);
+      bool has_bc = false;
+      for (std::size_t j = 0; j < dofs.size(); ++j)
+      {
+        if (has_bc)
+          break;
+        for (std::size_t k = 0; k < dofmap_bs; ++k)
+          if (dof_markers[dofs[j] * dofmap_bs + k])
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:179]: Missing `{`.

File: cpp/dolfinx/fem/DirichletBC.h [hunk 2]
```diff
@@ -293,6 +352,7 @@ class DirichletBC
         _dofs1_g(std::forward<typename U::value_type>(V_g_dofs[1])),
         _owned_indices0(num_owned(*_function_space, _dofs0))
   {
+    _marker = mark_cells(*_function_space, _dofs0);
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:355]: Can't this go in the initialiser list?

File: cpp/dolfinx/fem/DirichletBC.h [hunk 3]
```diff
@@ -131,6 +131,62 @@ template <typename T>
 class DirichletBC
 {
 private:
+  /// @brief Mark cells with bc dofs
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:134]: Is there a reason why this is in the header file?

File: cpp/dolfinx/fem/DirichletBC.h [hunk 4]
```diff
@@ -452,6 +512,22 @@ class DirichletBC
     }
   }
 
+  /// Set markers[i] = true if cell i has a dof in with boundary condition
+  /// applied. Value of markers[i] is not changed otherwise.
+  /// @param[in,out] markers Entry makers[i] is set to true if cell i in
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:517]: Is it really an `in` arg? Do the input values get used?

File: cpp/dolfinx/fem/DirichletBC.h [hunk 5]
```diff
@@ -452,6 +512,22 @@ class DirichletBC
...
+  /// V0 has a dof with a boundary condition applied, i.e. dofs which are
+  /// fixed by a boundary condition. Other entries in @p markers are left
+  /// unchanged.
+  void mark_cells(const std::span<std::int8_t>& markers) const
+  {
+
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:523]: Remove space.

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 6]
```diff
@@ -54,7 +54,8 @@ void _lift_bc_cells(
     int cstride, const std::span<const std::uint32_t>& cell_info,
     const std::span<const T>& bc_values1,
     const std::span<const std::int8_t>& bc_markers1,
-    const std::span<const T>& x0, double scale)
+    const std::span<const std::int8_t>& bc_cells1, const std::span<const T>& x0,
```
Comment (garth-wells) [inline cpp/dolfinx/fem/assemble_vector_impl.h:57]: Pass spans by copy.

This function signature is a bit muddled now. We pass `cells` and `bc_cells1`. It would be neater to pass just the cells to be (1) assembled over and (2) which also have bcs applied.

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 7]
```diff
@@ -149,8 +119,6 @@ void _lift_bc_cells(
           {
             const T bc = bc_values1[jj];
             const T _x0 = x0.empty() ? 0.0 : x0[jj];
```
Comment (garth-wells) [inline cpp/dolfinx/fem/assemble_vector_impl.h:121]: Use `0` rather than `0.0` - the type `T` isn't known.

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 8]
```diff
@@ -149,8 +119,6 @@ void _lift_bc_cells(
...
-            // const T _x0 = 0.0;
-            // be -= Ae.col(bs1 * j + k) * scale * (bc - _x0);
```
Comment (garth-wells) [inline cpp/dolfinx/fem/assemble_vector_impl.h:153]: Might be helpful to leave this but with an improved comment - it shows more clearly what the below mathematical operation is.

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 9]
```diff
@@ -166,7 +134,6 @@ void _lift_bc_cells(
           {
             const T bc = bc_values1[jj];
             const T _x0 = x0.empty() ? 0.0 : x0[jj];
-            // be -= Ae.col(bs1 * j + k) * scale * (bc - _x0);
```
Comment (garth-wells) [inline cpp/dolfinx/fem/assemble_vector_impl.h:169]: See comment above.

File: cpp/dolfinx/fem/DirichletBC.h [hunk 10]
```diff
@@ -131,6 +131,62 @@ template <typename T>
 class DirichletBC
 {
 private:
+  /// @brief Mark cells with bc dofs
+  ///
+  /// Returns an array of size number of local cells, where 1 indicates that the
+  /// cell has a dof from the incoming list in the dofmap
+  /// @param[in] V The function space
+  /// @param[in] dofs The degrees of freedom (local indices, unrolled),
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:139]: Which degrees of freedom? Doc string could be clearer.

File: cpp/dolfinx/fem/assemble_vector_impl.h (hunk 6 above)
Comment (jorgensd) [reply cpp/dolfinx/fem/assemble_vector_impl.h:57]: > Pass spans by copy.

Should I change this for `bc_values1`, `bc_markers1` as well?

File: cpp/dolfinx/fem/assemble_vector_impl.h (hunk 8 above)
Comment (jorgensd) [reply cpp/dolfinx/fem/assemble_vector_impl.h:153]: I've left it as it was, as I don't know a nicer way of write column of Ae, except:
```c++
// Ae[:, bs1 * j + k]
```

File: cpp/dolfinx/fem/DirichletBC.h (hunk 10 above)
Comment (jorgensd) [reply cpp/dolfinx/fem/DirichletBC.h]: I've tried to clarify this now by renaming the convenience function and giving it proper documentation.

**Comment by jorgensd** [comment]
Slightly modified plotting and run script:
```python
import dolfinx
from mpi4py import MPI
import time
import ufl
import numpy as np

import os
import os.path

Ns = [25, 50, 75]
Ps = [1,2,3]
num_runs = 10

for P in Ps:
    name = f"results_{P}.csv"

    file_exists = os.path.exists(f"{os.getcwd()}/{name}")
    if file_exists:
        outfile = open(name, "a")
    else:
        outfile = open(name, "w")
        print("Hash Dofs Time", file=outfile)


    for N in Ns:
        mesh = dolfinx.mesh.create_unit_cube(MPI.COMM_WORLD, N, N, N)

        V = dolfinx.fem.FunctionSpace(mesh, ("Lagrange", P))
        u = ufl.TrialFunction(V)
        v = ufl.TestFunction(V)
        f = dolfinx.fem.Function(V)
        f.interpolate(lambda x: np.sin(x[0]))

        u_bc = dolfinx.fem.Function(V)
        u_bc.interpolate(lambda x: np.ones(x.shape[1])+np.cos(x[0]))
        mesh.topology.create_connectivity(mesh.topology.dim-1, mesh.topology.dim)
        exterior_facets = dolfinx.mesh.exterior_facet_indices(mesh.topology)
        dofs = dolfinx.fem.locate_dofs_topological(
            V, mesh.topology.dim-1, exterior_facets)

        bc = dolfinx.fem.dirichletbc(u_bc, dofs)

        a = ufl.inner(u, v) * ufl.dx
        L = ufl.inner(f, v) * ufl.dx

        af = dolfinx.fem.form(
            a,  jit_options={"cffi_extra_compile_args": ["-march=native", "-O3"]})
        Lf = dolfinx.fem.form(
            L, jit_options={"cffi_extra_compile_args": ["-march=native", "-O3"]})

        
        b = dolfinx.fem.petsc.assemble_vector(Lf)  
        num_dofs = V.dofmap.index_map.size_global * V.dofmap.index_map_bs

        for i in range(num_runs):
            start = time.perf_counter()
            dolfinx.fem.petsc.apply_lifting(b, [af], [[bc]])
            end = time.perf_counter()
            print(f"{dolfinx.git_commit_hash} {num_dofs} {end-start}", file=outfile)

outfile.close()
```
```python
import matplotlib.pyplot as plt
import seaborn
import pandas

Ps = [1, 2, 3]
for P in Ps:

    df = pandas.read_csv(f"results_{P}.csv", delimiter=" ")

    plot = seaborn.catplot(data=df, kind="swarm", x="Dofs",
                        y="Time", hue="Hash")
    plot.set(yscale="log")
    plt.grid()
    plt.savefig(f"runtimes_{P}.png")
```
with results 
P1
P2
P3
Looking at the raw data, we have that a run on main (with P1) takes `0.01926 s` with `438976` dofs, while this branch is at `0.01395 s`

File: cpp/dolfinx/fem/DirichletBC.h [hunk 11]
```diff
@@ -451,7 +516,23 @@ class DirichletBC
     }
   }
 
-  /// Set markers[i] = true if dof i has a boundary condition applied.
+  /// Set markers[i] = true if cell i has a dof in with boundary condition
+  /// applied. Value of markers[i] is not changed otherwise.
+  /// @param[in,out] markers Entry makers[i] is set to true if cell i in
```
Comment (jhale) [inline cpp/dolfinx/fem/DirichletBC.h:521]: typo makers

**Review by jhale — APPROVED** [review]

File: cpp/dolfinx/fem/DirichletBC.h (hunk 11 above)
Comment (jorgensd) [inline cpp/dolfinx/fem/DirichletBC.h:521]:
```suggestion
  /// @param[in,out] markers Entry markers[i] is set to true if cell i in
```
Fix typo

File: cpp/dolfinx/fem/DirichletBC.h [hunk 12]
```diff
@@ -451,7 +516,23 @@ class DirichletBC
...
-  /// Set markers[i] = true if dof i has a boundary condition applied.
+  /// Set markers[i] = true if cell i has a dof in with boundary condition
+  /// applied. Value of markers[i] is not changed otherwise.
+  /// @param[in,out] markers Entry markers[i] is set to true if cell i in
+  /// V0 has a dof with a boundary condition applied, i.e. dofs which are
+  /// fixed by a boundary condition. Other entries in @p markers are left
+  /// unchanged.
+  void mark_cells(const std::span<std::int8_t>& markers) const
+  {
+
```
Comment (jorgensd) [inline cpp/dolfinx/fem/DirichletBC.h:527]:
```suggestion
```

**Comment by garth-wells** [comment]
I have some more comments on this PR, but need some more time.

**Comment by jorgensd** [comment]
@garth-wells  Is there anything specific missing from this PR?

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 13]
```diff
@@ -26,6 +26,43 @@
 namespace dolfinx::fem::impl
 {
 
+/// Extract a sub-set of entities whose cell index (local to process) is marked
+///
+/// @param[in] entities The entities.
+/// @param[in] cell_indicator An indicator vector. The size of cell_indicator
```
Comment (garth-wells) [inline cpp/dolfinx/fem/assemble_vector_impl.h:32]: This doesn't say what constitutes being 'marked'.

**Review by garth-wells — CHANGES_REQUESTED** [review]
First pass. Second pass will follow clarification of some points.

File: cpp/dolfinx/fem/DirichletBC.h [hunk 14]
```diff
@@ -132,6 +132,70 @@ template <typename T>
...
+    std::vector<std::int32_t> marker;
+    marker.reserve(num_cells);
+    for (std::size_t i = 0; i < num_cells; ++i)
+    {
+      auto dofs = dofmap->cell_dofs(i);
+      bool has_bc = false;
+      for (std::int32_t dof : dofs)
+      {
+        if (has_bc)
+          break;
```
[the remark this answers is not in the document]
Comment (jorgensd) [reply cpp/dolfinx/fem/DirichletBC.h:183]: We can remove the break statement, but that would mean that if the first dof finds a corresponding bc, we would loop through all other degrees of freedom and check them as well. I guess we could use `if (marker.back() == i)` if that helps the logic.

File: cpp/dolfinx/fem/DirichletBC.h [hunk 15]
```diff
@@ -448,7 +516,22 @@ class DirichletBC
...
   }
 
-  /// Set markers[i] = true if dof i has a boundary condition applied.
+  /// Set markers[i] = true if cell i has a dof in with boundary condition
+  /// applied. Value of markers[i] is not changed otherwise.
+  /// @param[in,out] markers Entry markers[i] is set to true if cell i in
+  /// V0 has a dof with a boundary condition applied, i.e. dofs which are
+  /// fixed by a boundary condition. Other entries in @p markers are left
+  /// unchanged.
+  void mark_cells(const std::span<std::int8_t>& markers) const
```
[the remark this answers is not in the document]
Comment (jorgensd) [reply cpp/dolfinx/fem/DirichletBC.h:525]: When one has multiple Dirichlet conditions, you just want to mark all cells that have such a condition. You do not want to merge and sort N sorted arrays to get the set of cells you need.

File: cpp/dolfinx/fem/DirichletBC.h (hunk 14 above)
[the remark this answers is not in the document]
Comment (garth-wells) [reply cpp/dolfinx/fem/DirichletBC.h:183]: The point isn't that the break needs to be removed, but that having a break at the start of a loop rather than the end is strange and hard to follow.

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 16]
```diff
@@ -26,6 +26,43 @@
...
+template <std::size_t stride>
+std::vector<std::int32_t>
+_extract_entities(std::span<const std::int32_t> entities,
+                  std::span<const std::int8_t> cell_indicator)
+{
+  std::vector<std::int32_t> sub_entities;
+  sub_entities.reserve(entities.size());
+  for (std::size_t e = 0; e < entities.size(); e += stride)
+  {
+    if constexpr (stride == 1)
```
[the remark this answers is not in the document]
Comment (jorgensd) [reply cpp/dolfinx/fem/assemble_vector_impl.h:46]: And it would never happen, as we only have three types of entities, cells (stride 1), exterior facets (stride 2), interior facets (stride 4), as mentioned in the `@tparam` documentation.

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 17]
```diff
@@ -26,6 +26,43 @@
...
+    {
+      std::copy_n(std::next(entities.begin(), e), stride,
+                  std::back_inserter(sub_entities));
+    }
+    else if constexpr (stride == 4)
+    {
+      if (cell_indicator[entities[e]] or cell_indicator[entities[e + 2]])
+        std::copy_n(std::next(entities.begin(), e), stride,
+                    std::back_inserter(sub_entities));
+    }
```
Comment (jorgensd) [inline cpp/dolfinx/fem/assemble_vector_impl.h:61]:
```suggestion
      if (cell_indicator[entities[e]] or cell_indicator[entities[e + 2]])
      {
        std::copy_n(std::next(entities.begin(), e), stride,
                    std::back_inserter(sub_entities));
      }
    }
```

File: cpp/dolfinx/fem/assemble_vector_impl.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (garth-wells) [reply cpp/dolfinx/fem/assemble_vector_impl.h:46]: The documentation doesn't impose any constraints. Constraints should be checked in the code when possible. 

In any case, `constexpr` seems like an unnecessary complication. It would be simpler to handle each case separately.

**Comment by jorgensd** [comment]
> I have some more comments on this PR, but need some more time.

I think we should not use the function `_extract_entities`, as it is more costly than having an if-test inside assembly with the cell indicator.

If we send in `cell_indicator` as a reference, and the full set of integration entities, we only have one 
```c++
if !(cell_indicator[cell_Index])
    pass;
```
inside `assemble_cells` and `assemble_exterior_facets`, while we would need
```c++
if (!cell_indicator[cell0] and !cell_indicator[cell1])
    pass; 
```
compared to the current strategy:
1. Create a new vector which has `num_entities*num_entities_per_cell` reserved in memory.
2. Check the same if statement as above
3. Copy data if check is true

@jhale @IgorBaratta @chrisrichardson any thoughts on this?

**Review by IgorBaratta — APPROVED** [review]

File: cpp/dolfinx/fem/DirichletBC.h [hunk 18]
```diff
@@ -132,6 +132,71 @@ template <typename T>
 class DirichletBC
 {
 private:
+  /// @brief Build list of cells (local to process, including ghosts)
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:135]: I don't think we need 'local to the process'. It's implicit throughout the code.

File: cpp/dolfinx/fem/DirichletBC.h [hunk 19]
```diff
@@ -132,6 +132,71 @@ template <typename T>
...
+  /// which contain any of the input degrees of freedom.
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:136]: This is somewhat ambiguous - elsewhere we only associate interior dofs with the 'cell'. I presume you mean cell closure?

File: cpp/dolfinx/fem/DirichletBC.h [hunk 20]
```diff
@@ -132,6 +132,71 @@ template <typename T>
...
+  ///
+  /// Given a function space `V` and a set of `dofs`, return cell `c` if
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:138]: The function arguments give the type, so no need to repeat. Can this text be removed and covered by the parameter and return docstrings?

File: cpp/dolfinx/fem/DirichletBC.h [hunk 21]
```diff
@@ -132,6 +132,71 @@ template <typename T>
...
+  /// `V.dofmap()->cell_dofs(i)` contains any of the entries in `dofs`.
+  ///
+  /// @param[in] V The function space
+  /// @param[in] dofs A list of degrees of freedom (local indices, unrolled),
+  /// including ghosts
+  /// @returns List of cell indices (local to process, including ghosts) of
+  /// which has any of the input `dofs`
+  std::vector<std::int32_t> locate_cells(const FunctionSpace& V,
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:146]: Should this function be static? Or a free function in the `impl` namespace? It's not obvious that it needs to be a non-const member.

File: cpp/dolfinx/fem/DirichletBC.h [hunk 22]
```diff
@@ -448,7 +517,22 @@ class DirichletBC
...
   }
 
-  /// Set markers[i] = true if dof i has a boundary condition applied.
+  /// Set markers[i] = 1 if cell i has a dof in with boundary condition
+  /// applied. Value of markers[i] is not changed otherwise.
+  /// @param[in,out] markers Entry markers[i] is set to true if cell i in
+  /// V0 has a dof with a boundary condition applied, i.e. dofs which are
+  /// fixed by a boundary condition. Other entries in @p markers are left
+  /// unchanged.
+  void mark_cells(const std::span<std::int8_t>& markers) const
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:526]: Pass by value rather than reference.

File: cpp/dolfinx/fem/DirichletBC.h [hunk 23]
```diff
@@ -448,7 +517,22 @@ class DirichletBC
     }
   }
 
-  /// Set markers[i] = true if dof i has a boundary condition applied.
+  /// Set markers[i] = 1 if cell i has a dof in with boundary condition
+  /// applied. Value of markers[i] is not changed otherwise.
+  /// @param[in,out] markers Entry markers[i] is set to true if cell i in
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:522]: What does `Entry` mean?

What are the preconditions on `markers`? Use the doxygen syntax for preconditions to document.

File: cpp/dolfinx/fem/DirichletBC.h [hunk 24]
```diff
@@ -448,7 +517,22 @@ class DirichletBC
...
+  /// V0 has a dof with a boundary condition applied, i.e. dofs which are
```
Comment (garth-wells) [inline cpp/dolfinx/fem/DirichletBC.h:523]: What is `V0`?

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 25]
```diff
@@ -26,6 +26,56 @@
 namespace dolfinx::fem::impl
 {
 
+/// Extract a sub-set of entities whose cell index (local to process) is marked
```
Comment (garth-wells) [inline cpp/dolfinx/fem/assemble_vector_impl.h:29]: I don't follow this fully.

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 26]
```diff
@@ -26,6 +26,56 @@
...
+///
+/// @param[in] entities The entities.
+/// @param[in] cell_indicator An indicator vector. The size of cell_indicator
```
Comment (garth-wells) [inline cpp/dolfinx/fem/assemble_vector_impl.h:32]: 'An indicator vector' doesn't say much.

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 27]
```diff
@@ -26,6 +26,56 @@
...
+/// should be smaller than max(entities). A cell is marked if it contains a
+/// non-zero value.
+/// @tparam stride  The number of entries in each entity (1 for cells, 2 for
```
Comment (garth-wells) [inline cpp/dolfinx/fem/assemble_vector_impl.h:35]: 'The number of entries in each entity'?

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 28]
```diff
@@ -26,6 +26,56 @@
...
+/// exterior facets, 4 for interior facets)
+/// @returns A list of sub-entities
+template <std::size_t stride>
+std::vector<std::int32_t>
+_extract_entities(std::span<const std::int32_t> entities,
+                  std::span<const std::int8_t> cell_indicator)
+{
+  std::vector<std::int32_t> sub_entities;
+  sub_entities.reserve(entities.size());
+  if (stride == 1)
```
Comment (garth-wells) [inline cpp/dolfinx/fem/assemble_vector_impl.h:45]: Should this be `constexpr`?

Is there a measurable performance benefit to templating over `stride`?

File: cpp/dolfinx/fem/assemble_vector_impl.h [hunk 29]
```diff
@@ -727,6 +695,7 @@ void assemble_interior_facets(
 /// @param[in] bc_values1 The boundary condition 'values'
 /// @param[in] bc_markers1 The indices (columns of A, rows of x) to
 /// which bcs belong
+/// @param[in] bc_cells1 The cells which are associated with any bc dof
```
Comment (garth-wells) [inline cpp/dolfinx/fem/assemble_vector_impl.h:698]: 'Cells to which a Dirichlet boundary condition is applied.`

**Review by garth-wells — CHANGES_REQUESTED** [review]

File: cpp/dolfinx/fem/assemble_vector_impl.h (hunk 28 above)
Comment (jorgensd) [reply cpp/dolfinx/fem/assemble_vector_impl.h:45]: This was a const-expr: https://github.com/FEniCS/dolfinx/pull/2395#discussion_r1023622907, but you asked me to remove this.

Comment (garth-wells) [reply cpp/dolfinx/fem/assemble_vector_impl.h:45]: Yes, but the template parameter is still there. Can it be done simply without a template parameter and without special cases?

**Comment by garth-wells** [comment]
Closing as it's been open for a long time. Re-open if still of interest.
